src/main/java/winstone/Launcher.java

```diff
@@ -229,9 +231,19 @@ private void writePortToFileIfNeeded() throws IOException {
                     Path portFile = Paths.get(portFileName);
                     Path portDir = portFile.getParent();
                     Files.createDirectories(portDir);
-                    try (BufferedWriter writer = Files.newBufferedWriter(portFile, StandardCharsets.UTF_8)) {
+                    Path tmpPath = Files.createTempFile(portDir, "temp", "");
+                    try (BufferedWriter writer = Files.newBufferedWriter(tmpPath, StandardCharsets.UTF_8)) {
                         writer.write(Integer.toString(port));
                     }
+
+                    try {
+                        Files.move(tmpPath, portFile, StandardCopyOption.ATOMIC_MOVE);
+                    } catch (AtomicMoveNotSupportedException e) {
+                        Logger.logDirectMessage(Level.WARNING, "", "Atomic move does not supported", e);
+                        Files.move(tmpPath, portFile, StandardCopyOption.REPLACE_EXISTING);
```

Comment:
If we encounter an exception on the second attempt, we will lose the stack trace from the first attempt, which is valuable for debugging. Better to attach it as a suppressed exception, which preserves the original stack trace from the first attempt:

```suggestion
                        try {
                            Files.move(tmpPath, portFile, StandardCopyOption.REPLACE_EXISTING);
                        } catch (IOException e2) {
                            e2.addSuppressed(e);
                            throw e2;
                        }
```